### ISSUE TITLE
Allowing escaped placeholders in template strings when using rule "quotes" with option "single" or "double"

### ISSUE THREAD

**Issue by SebastianG77** [eslint/eslint#12466]
**What rule do you want to change?**

[quotes](https://eslint.org/docs/rules/quotes)

**Does this change cause the rule to produce more or fewer warnings?**

fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option or new default behavior

**Please provide some example code that this change will affect:**

```js
const foo = `\${package.version}`
```

**What does the rule currently do for this code?**

When activating [quotes](https://eslint.org/docs/rules/quotes)  with option "double" or "single" as well as [no-template-curly-in-string](https://eslint.org/docs/rules/no-template-curly-in-string), it currently is impossible to build a string that interprets the string representation of a placeholder (i.e. '${package.version}' without either violating [quotes](https://eslint.org/docs/rules/quotes) or [no-template-curly-in-string](https://eslint.org/docs/rules/no-template-curly-in-string) unless using ugly workarounds like ``const foo = `${''}\${package.version}``.

**What will the rule do after it's changed?**

It will allow building strings that contain a string representation of placeholder strings when using [quotes](https://eslint.org/docs/rules/quotes) and [no-template-curly-in-string](https://eslint.org/docs/rules/no-template-curly-in-string) simultaneously, since then template strings are also valid according to the less strict handling of [quotes](https://eslint.org/docs/rules/quotes) if they do not contain a placeholder but its escaped string representation.

Since this issue first came up in [StandardJS](https://github.com/standard/standard) please also refer to ["Unexpected template string expression" vs. "Strings must use singlequote" #1437](https://github.com/standard/standard/issues/1437) for further information.

**Are you willing to submit a pull request to implement this change?**

Yes

**Comment by mdjermanovic** [comment]
Hi @SebastianG77, thanks for the issue!

In my opinion, this is a case for an `eslint-disable` comment.

**Comment by SebastianG77** [comment]
Using eslint-disable would also just be a workaround for me and I would have to adjust every single comment that contains the string representation of a placeholder. It certainly will be more convenient and cleaner if there is at least an option that completely avoids this problem for all strings. Will appreciate if we can find a cleaner solution for this problem.

**Comment by mdjermanovic** [comment]
It's just my personal opinion, other team members may disagree. If there are too many cases like this it might make sense to disable the rule at the top of the file or in the whole project.

I think it was also the recommended way to deal with this situation when the rule was included in the standard: [standard/standard#594 (comment)](https://github.com/standard/standard/issues/594#issuecomment-242256610)

**Comment by mdjermanovic** [comment]
To avoid confusion, I meant to disable `no-template-curly-in-string`, not `quotes`.

**Comment by mightyiam** [comment]
Just a reminder here that issue author is willing to submit a PR. Pending further input from maintainers.

**Comment by kaicataldo** [comment]
> It will allow building strings that contain a string representation of placeholder strings __when using quotes and no-template-curly-in-string simultaneously__, since then template strings are also valid according to the less strict handling of quotes if they do not contain a placeholder but its escaped string representation.

Emphasis above is mine. I don't think we'd be able to support this, because ESLint rules do not know about each other by design. I agree with @mdjermanovic that `eslint-disable-*` comments are the best solution to this issue right now.

**Comment by mightyiam** [comment]
To summarize, what @SebastianG77 is suggesting is that
```
`\${foo}`
```
Is allowed when `quotes: ['single']` or `quotes: ['double']`.

Is this correct, @SebastianG77 ?

This would be a non-breaking slight relaxing of the `quotes` option. Am I right?

**Comment by SebastianG77** [comment]
@kaicataldo 
Are there more rules in eslint that are contradictorily? To me it looks like a bug, if two rules are so contradictorily that there is no possibility to write certain code in any way. I also do not think it is good style to use comments as the only option to allow certain string values, because comments should usually be used to comment code and not to deactivate linter rules.

@mightyiam 
yes, that is correct. Ideally, it should always be possible to use `` `\${foo}` ``, irrespective of the chosen `quotes` option. As a result, `no-template-curly-in-string` and `quotes` do not need to know of each other, since the rule `quotes` will just be soften up a bit.

**Comment by kaicataldo** [comment]
I see - I misunderstood the proposal. Can you explain why the [allowTemplateLiterals](https://eslint.org/docs/rules/quotes#allowtemplateliterals) option in the `quotes` rule doesn't solve this issue?

**Comment by SebastianG77** [comment]
I just tried out `allowTemplateLiterals`. The problem with this option is you can always use backticks resulting in a rule that is too soft. I am looking for a possibility to force developers using single quotes unless there is a good reason not to do it like i.e. using template strings or if it is simply not possible under some circumstances to write a string with single quotes (as when trying to write the string `'${foo}'` while also using rule `no-template-curly-in-string`). While usually `quotes` without option `allowTemplateLiterals` does it quiet good in most cases, I am still looking for just a slightly less strict rule that avoids the contradiction when also using  `no-template-curly-in-string`.

**Comment by kaicataldo** [comment]
Thank you for providing more context around the problem you're trying to solve.

So, just to clarify: the proposal here is to add an option to `quotes` to allow for the use of backticks when `quotes` is configured with `'single'` or `'double'` and the string contains `\${}`, correct?

**Comment by SebastianG77** [comment]
Exactly. 

It could be either an option or default behavior of `quotes`. I would be fine with any of these solutions.

**Comment by kaicataldo** [comment]
Thanks for clarifying! I personally think this a good case for an `eslint-disable-*` comment for the `no-template-curly-in-string` rule, but let’s see what the rest of the team thinks.

**Comment by mightyiam** [comment]
@eslintbot thank you for reminding us about this. We'd like some more team members to take a peek and at least one to re-open this issue.
